CMakeLists.txt

```diff
@@ -207,22 +207,40 @@ endif()
 
 # If available and wanted, use compiler cache.
 if(LIBCELLML_COMPILER_CACHE)
-  if(WIN32)
-    if(CLCACHE_EXE)
-      set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
-      execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
-                      RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
-      if(RESULT EQUAL 0)
-        set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
-      endif()
-    endif()
+  if(BUILDCACHE_EXE)
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${BUILDCACHE_EXE})
   else()
-    if(CCACHE_EXE)
-      set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+    if(WIN32)
+      if(CLCACHE_EXE)
+        set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
+        execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
+                        RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
+        if(RESULT EQUAL 0)
+          set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
+        endif()
+      endif()
+    else()
+      if(CCACHE_EXE)
+        set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+      endif()
     endif()
   endif()
 endif()
 
+# On Windows, if we are using MSVC then replace some compiler flags.
+if(WIN32 AND MSVC)
+  # If we are treating warnings as errors then remove the /W3 option since it 
+  # will eventually be replaced with the /W4 and /WX options.
```

Comment:
struggling to understand this logic. If it will eventually be replaced by other options why remove it here?

Reply:
Because we add `/W4 /WX` to a target using `target_compile_options()` and this doesn't allow to remove `/W3`. So, what we are doing here is to remove `/W3` all possible targets.